File: 001/D/main.rb
#! ruby
# -*- coding: utf-8 -*-

lambda{
	# データ読み込み
	n = gets.to_i
	start, goal = *gets.split(/ /).map{|c| c.to_f}

	ranges = Array.new(n+1) do
		l, r = *gets.split(/ /).map{|c| c.to_f}
		{ :l => l, :r => r }
	end
	# Goalの範囲はgoal地点のみとみなす
	ranges[n][:l] = ranges[n][:r] = goal

	# 左右の視野角(簡易)
	l_view_angle = -Float::MAX
	r_view_angle = Float::MAX
	# 左右障害になりうる壁のy座標
	l_wall_y = r_wall_y = 0

	dist = 0.0
	turning_y = 0
	y = 1
	while y <= n do
		rng = ranges[y]

		l_va = (rng[:l] - start) / (y - turning_y)
		if l_va >= l_view_angle
			l_view_angle = l_va
			l_wall_y = y
		end
		r_va = (rng[:r] - start) / (y - turning_y)
		if r_va <= r_view_angle
			r_view_angle = r_va
			r_wall_y = y
		end

		# 左の壁が右の壁の右方にきた
		if l_view_angle > r_view_angle
			if y == r_wall_y # 右の壁が奥=左カーブ
				x = ranges[l_wall_y][:l]
				y = l_wall_y
			else
				x = ranges[r_wall_y][:r]
				y = r_wall_y
			end
			dist += Math.hypot(x - start, y - turning_y)
			start, turning_y = x, y
			l_view_angle = -Float::MAX
			r_view_angle = Float::MAX
		end
		y += 1
	end
	dist += Math.hypot(goal - start, n - turning_y)
	puts dist
}.call

File: 001/D/main.cpp
#include <cstdio>
#include <cfloat>
#include <cmath>
#include <iostream>
#include <vector>

struct Range
{
	long l;
	long r;
	Range(long l, long r){ this->l = l; this->r = r; }
};

int main()
{
	long n, start, goal;
	std::cin >> n >> start >> goal;

	std::vector<Range> ranges;
	ranges.reserve(n+1);
	for (long i = 0; i <= n; ++i) {
		long l, r;
		std::cin >> l >> r;
		ranges.push_back(Range(l,r));
	}
	ranges[n] = (Range(goal,goal));

	double l_view_angle = -DBL_MAX;
	double r_view_angle = DBL_MAX;
	long l_wall_y = 0, r_wall_y = 0;
	double dist = 0;
	long turning_y = 0;
	for (long y = 1; y <= n; ++y) {
		const Range rng = ranges[y];
		double l_va = (double)(rng.l - start) / (y - turning_y);
		if (l_va > l_view_angle) {
			l_view_angle = l_va;
			l_wall_y = y;
		}
		double r_va = (double)(rng.r - start) / (y - turning_y);
		if (r_va < r_view_angle) {
			r_view_angle = r_va;
			r_wall_y = y;
		}
		if (l_view_angle > r_view_angle) {
			long x;
			if (y == r_wall_y) {
				x = ranges[l_wall_y].l;
				y = l_wall_y;
			} else {
				x = ranges[r_wall_y].r;
				y = r_wall_y;
			}
			dist += hypot(x - start, y - turning_y);
			start = x; turning_y = y;
			l_view_angle = -DBL_MAX;
			r_view_angle = DBL_MAX;
		}
	}
	dist += hypot(goal - start, n - turning_y);
	printf("%.12lf\n", dist);
	return 0;
}
